### PR TITLE
Avoid frequent malloc in measure and U3

### PR DESCRIPTION
**Pull request by hhorii**
### Summary
This PR reduces numbers of calls to perform measure and U3 in statevector

### Details and comments
The original `apply_gate_mcu3` inefficiently generates a matrix for qubitvector while generating a `cmatrix` object. Also, `probabilities` in qubitvector is not specialized for single-qubit parameter. This PR fixes these issues.

### CODE REVIEW

File: src/framework/utils.hpp [hunk 1]
```diff
@@ -335,7 +335,143 @@ cmatrix_t Matrix::U3(double theta, double phi, double lambda) {
...
+class VMatrix {
+public:
+  // Single-qubit gates
+  const static cvector_t I;     // name: "id"
+  const static cvector_t X;     // name: "x"
+  const static cvector_t Y;     // name: "y"
+  const static cvector_t Z;     // name: "z"
+  const static cvector_t H;     // name: "h"
+  const static cvector_t S;     // name: "s"
+  const static cvector_t Sdg;   // name: "sdg"
```
Comment (atilag) [inline src/framework/utils.hpp]: To keep consistency with the rest of constant names, I'd rename it to: `SDG`

File: src/framework/utils.hpp [hunk 2]
```diff
@@ -335,7 +335,143 @@ cmatrix_t Matrix::U3(double theta, double phi, double lambda) {
...
+  const static cvector_t T;     // name: "t"
+  const static cvector_t Tdg;   // name: "tdg"
```
Comment (atilag) [inline src/framework/utils.hpp]: `Tdg` -> `TDG`

File: src/framework/utils.hpp [hunk 3]
```diff
@@ -335,7 +335,143 @@ cmatrix_t Matrix::U3(double theta, double phi, double lambda) {
...
+
+  // Two-qubit gates
+  const static cvector_t CX;    // name: "cx"
+  const static cvector_t CZ;    // name: "cz"
+  const static cvector_t SWAP;  // name: "swap"
+  const static cvector_t CR;    // TODO
+  const static cvector_t CR90;  // TODO
+
+  // Identity Matrix
+  static cvector_t Identity(size_t dim);
```
Comment (atilag) [inline src/framework/utils.hpp]: This is `const` and we use `lower_snake_case` for method names:
```c++
static cvector_t identity() const;
```

File: src/framework/utils.hpp [hunk 4]
```diff
@@ -335,7 +335,143 @@ cmatrix_t Matrix::U3(double theta, double phi, double lambda) {
...
+
+  // Single-qubit waltz gates
+  static cvector_t U1(double lam);
```
Comment (atilag) [inline src/framework/utils.hpp]: `const` and lower case:
```c++
static cvector_t u1(double lam) const;
```

File: src/framework/utils.hpp [hunk 5]
```diff
@@ -335,7 +335,143 @@ cmatrix_t Matrix::U3(double theta, double phi, double lambda) {
...
+  static cvector_t U2(double phi, double lam);
```
Comment (atilag) [inline src/framework/utils.hpp]: `const` and lower case:
```c++
static cvector_t u2(...) const;
```

File: src/framework/utils.hpp [hunk 6]
```diff
@@ -335,7 +335,143 @@ cmatrix_t Matrix::U3(double theta, double phi, double lambda) {
...
+  static cvector_t U3(double theta, double phi, double lam);
```
Comment (atilag) [inline src/framework/utils.hpp]: `const` and lower case:
```c++
static cvector_t u3(...) const;
```

File: src/framework/utils.hpp [hunk 7]
```diff
@@ -335,7 +335,143 @@ cmatrix_t Matrix::U3(double theta, double phi, double lambda) {
...
+
+  // Complex arguments are implemented by taking std::real
+  // of the input
+  inline static cvector_t U1(complex_t lam) {return U1(std::real(lam));}
```
Comment (atilag) [inline src/framework/utils.hpp]: `const` and lower case. There's no need to specify `inline` here as we are in a header file, so this is always inlined.
```c++
static cvector_t u1(...) const;
```

File: src/framework/utils.hpp [hunk 8]
```diff
@@ -335,7 +335,143 @@ cmatrix_t Matrix::U3(double theta, double phi, double lambda) {
...
+  inline static cvector_t U2(complex_t phi, complex_t lam) {
```
Comment (atilag) [inline src/framework/utils.hpp]: * Remove `inline`
* snake_lower_case => `u2(...)`
* `const`

File: src/framework/utils.hpp [hunk 9]
```diff
@@ -335,7 +335,143 @@ cmatrix_t Matrix::U3(double theta, double phi, double lambda) {
...
+    return U2(std::real(phi), std::real(lam));
+  }
+  inline static cvector_t U3(complex_t theta, complex_t phi, complex_t lam) {
```
Comment (atilag) [inline src/framework/utils.hpp]: * Remove `inline`
* snake_lower_case => `u3(...)`
* `const`

File: src/framework/utils.hpp [hunk 10]
```diff
@@ -335,7 +335,143 @@ cmatrix_t Matrix::U3(double theta, double phi, double lambda) {
...
+    return U3(std::real(theta), std::real(phi), std::real(lam));
+  };
+
+  // Return the matrix for a named matrix string
+  // Allowed names correspond to all the const static single-qubit
+  // and two-qubit gate members
+  inline static const cvector_t from_name(const std::string &name) {
```
Comment (atilag) [inline src/framework/utils.hpp]: * Remove `inline`
* `const`

File: src/framework/utils.hpp [hunk 11]
```diff
@@ -335,7 +335,143 @@ cmatrix_t Matrix::U3(double theta, double phi, double lambda) {
...
+    return *label_map_.at(name);
+  }
+
+  // Check if the input name string is allowed
+  inline static bool allowed_name(const std::string &name) {
```
Comment (atilag) [inline src/framework/utils.hpp]: * Remove `inline`
* `const`
* The convention for functions returning booleans is using the `is_xxxx()` semantic. So:
```c++
static bool is_allowed_named(...) const {
```

File: src/simulators/statevector/qubitvector.hpp [hunk 12]
```diff
@@ -772,20 +772,27 @@ void QubitVector<data_t>::zero() {
...
+  if (data_) {
+    if (prev_num_qubits != num_qubits_) {
+      free(data_);
+      data_ = nullptr;
+    }
+  }
+
   // Allocate memory for new vector
-  data_ = reinterpret_cast<complex_t*>(malloc(sizeof(complex_t) * data_size_));
+  if (data_ == nullptr)
```
Comment (atilag) [inline src/simulators/statevector/qubitvector.hpp]: What if `data_` is not `nullptr`? Shall we reallocate memory to fit with `data_size_`? Are we supposing that `data_size_` hasn't changed so there's no real need for reallocation?

File: src/framework/utils.hpp (hunk 1 above)
Comment (hhorii) [reply src/framework/utils.hpp]: Member names of `AER::Utils::VMatrix` follow `AER::Utils::Matrix`. I will fix the both.

File: src/framework/utils.hpp (hunk 2 above)
Comment (hhorii) [reply src/framework/utils.hpp]: Member names of `AER::Utils::VMatrix` follow `AER::Utils::Matrix`. I will fix the both.

File: src/framework/utils.hpp (hunk 11 above)
Comment (hhorii) [reply src/framework/utils.hpp]: static method cannot have `const`.

File: src/simulators/statevector/qubitvector.hpp (hunk 12 above)
Comment (hhorii) [reply src/simulators/statevector/qubitvector.hpp]: `AER::Base::State` keeps `state_t qreg_`, which is a `QV::QubitVector`. `State` is reused in `run_circuit_with_noise` and `run_circuit_without_noise` in `AER::Simulator::QasmController` across each shot simulation.

**Pull request merged**